File: CS50X/Final Project/project C/project.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define FILENAME "employees.txt"

typedef struct
{
    int id;
    char name[100];
    char position[100];
    double salary;
} Employee;

void createDatabase();
void viewEmployee();
void addEmployee();
void editEmployee();
void deleteEmployee();
void pause();
void clearScreen();
void printTitle(const char *title);
void saveEmployees(Employee *employees, int count);
int loadEmployees(Employee **employees);
int getNextId(Employee *employees, int count);

int main()
{
    createDatabase();

    while (1)
    {
        clearScreen();
        printTitle("Welcome to employee's system");
        int option;

        printf("Main Menu\n1. View employees.\n2. Add employees.\n3. Update employees.\n4. Delete "
               "employees.\n5. Exit.\nEnter your choice: ");

        if (scanf("%d", &option) != 1)
        {
            printf("\nOption invalid. Please try again.\n");
            pause();
            while (getchar() != '\n')
                ;
            continue;
        }

        switch (option)
        {
            case 1:
                viewEmployee();
                pause();
                break;
            case 2:
                addEmployee();
                pause();
                break;
            case 3:
                editEmployee();
                pause();
                break;
            case 4:
                deleteEmployee();
                pause();
                break;
            case 5:
                clearScreen();
                printTitle("Thanks for using. Goodbye!");
                pause();
                clearScreen();
                exit(0);
            default:
                printf("Option invalid. Please try again.\n");
                pause();
                break;
        }
    }

    return 0;
}

void createDatabase()
{
    FILE *file = fopen(FILENAME, "a");
    if (file == NULL)
    {
        printf("Error creating database file.\n");
    }
    else
    {
        fclose(file);
    }
}

void viewEmployee()
{
    clearScreen();
    printTitle("View employees");

    Employee *employees = NULL;
    int count = loadEmployees(&employees);

    if (count == 0)
    {
        printf("No employees found.\n");
    }
    else
    {
        printf("ID\tName\t\tPosition\tSalary\n");
        printf("-------------------------------------------------\n");
        for (int i = 0; i < count; i++)
        {
            printf("%d\t%s\t\t%s\t\t%.2f\n", employees[i].id, employees[i].name,
                   employees[i].position, employees[i].salary);
        }
    }

    free(employees);
}

void addEmployee()
{
    clearScreen();
    printTitle("Add employee");

    Employee *employees = NULL;
    int count = loadEmployees(&employees);

    Employee newEmployee;
    newEmployee.id = getNextId(employees, count);

    printf("Employee's name: ");
    scanf("%s", newEmployee.name);
    printf("Employee's position: ");
    scanf("%s", newEmployee.position);
    printf("Employee's salary: ");
    if (scanf("%lf", &newEmployee.salary) != 1)
    {
        printf("\nData invalid.\n");
        free(employees);
        return;
    }

    employees = realloc(employees, sizeof(Employee) * (count + 1));
    employees[count] = newEmployee;
    saveEmployees(employees, count + 1);

    printf("\nEmployee added successfully.\n");
    free(employees);
}

void editEmployee()
{
    clearScreen();
    printTitle("Edit employee");

    Employee *employees = NULL;
    int count = loadEmployees(&employees);

    if (count == 0)
    {
        printf("No employees found.\n");
        return;
    }

    int employeeId;
    printf("Employee's Id: ");
    if (scanf("%d", &employeeId) != 1)
    {
        printf("\nData invalid.\n");
        free(employees);
        return;
    }

    for (int i = 0; i < count; i++)
    {
        if (employees[i].id == employeeId)
        {
            printf("Employee found.\nEmployee: %s, %s, %.2f\n", employees[i].name,
                   employees[i].position, employees[i].salary);
            printf("\nEnter new information (leave blank to keep current): \n");

            char newName[100], newPosition[100];
            double newSalary;

            printf("Enter new name: ");
            scanf("%s", newName);
            if (strlen(newName) > 0)
            {
                strcpy(employees[i].name, newName);
            }

            printf("Enter new position: ");
            scanf("%s", newPosition);
            if (strlen(newPosition) > 0)
            {
                strcpy(employees[i].position, newPosition);
            }

            printf("Enter new salary: ");
            if (scanf("%lf", &newSalary) == 1)
            {
                employees[i].salary = newSalary;
            }

            saveEmployees(employees, count);
            printf("\nEmployee information updated successfully.\n");
            free(employees);
            return;
        }
    }

    printf("\nEmployee not found.\n");
    free(employees);
}

void deleteEmployee()
{
    clearScreen();
    printTitle("Delete employee");

    Employee *employees = NULL;
    int count = loadEmployees(&employees);

    if (count == 0)
    {
        printf("No employees found.\n");
        return;
    }

    int employeeId;
    printf("Employee's Id: ");
    if (scanf("%d", &employeeId) != 1)
    {
        printf("\nData invalid.\n");
        free(employees);
        return;
    }

    for (int i = 0; i < count; i++)
    {
        if (employees[i].id == employeeId)
        {
            printf("Employee found.\nEmployee: %s, %s, %.2f\n", employees[i].name,
                   employees[i].position, employees[i].salary);
            char option;
            printf("\nAre you sure you want to delete it? (y/n): ");
            scanf(" %c", &option);

            if (option == 'y' || option == 'Y')
            {
                for (int j = i; j < count - 1; j++)
                {
                    employees[j] = employees[j + 1];
                }

                saveEmployees(employees, count - 1);
                printf("\nEmployee deleted successfully.\n");
                free(employees);
                return;
            }
            else
            {
                printf("\nDeletion cancelled.\n");
                free(employees);
                return;
            }
        }
    }

    printf("\nEmployee not found.\n");
    free(employees);
}

void pause()
{
    printf("Press Enter to continue...");
    while (getchar() != '\n')
        ;
    getchar();
}

void clearScreen()
{
    printf("\033[H\033[J");
}

void printTitle(const char *title)
{
    printf("\n%s\n", title);
}

void saveEmployees(Employee *employees, int count)
{
    FILE *file = fopen(FILENAME, "w");
    if (file == NULL)
    {
        printf("Error opening file for writing.\n");
        return;
    }

    for (int i = 0; i < count; i++)
    {
        fprintf(file, "%d %s %s %.2f\n", employees[i].id, employees[i].name, employees[i].position,
                employees[i].salary);
    }

    fclose(file);
}

int loadEmployees(Employee **employees)
{
    FILE *file = fopen(FILENAME, "r");
    if (file == NULL)
    {
        *employees = NULL;
        return 0;
    }

    int count = 0;
    Employee temp;
    while (fscanf(file, "%d %s %s %lf", &temp.id, temp.name, temp.position, &temp.salary) == 4)
    {
        *employees = realloc(*employees, sizeof(Employee) * (count + 1));
        (*employees)[count++] = temp;
    }

    fclose(file);
    return count;
}

int getNextId(Employee *employees, int count)
{
    int maxId = 0;
    for (int i = 0; i < count; i++)
    {
        if (employees[i].id > maxId)
        {
            maxId = employees[i].id;
        }
    }

    return maxId + 1;
}

File: CS50X/Final Project/project Python/project.py
import os
import sqlite3
from tabulate import tabulate
from pyfiglet import Figlet

def main():
    createDatabase()
    while True:
        os.system('clear')
        figlet("Welcome to employee's system")
        try:
            option = int(input("Main Menu\n1. View employees.\n2. Add employees.\n3. Update employees.\n4. Delete employees.\n5. Exit.\nEnter your choice: "))

            match option:
                case 1:
                    viewEmployee()
                    pause()
                case 2:
                    addEmployee()
                    pause()
                case 3:
                    editEmployee()
                    pause()
                case 4:
                    deleteEmployee()
                    pause()
                case 5:
                    os.system('clear')
                    figlet("Thanks for using. Goodbyt!")
                    pause()
                    os.system('clear')
                    break
                case _:
                    print("Option invalid. Please try again.")
                    pause()
        except ValueError:
            print("\nOption invalid. Please try again.")
            pause()

def addEmployee():
    os.system('clear')
    figlet("Add employee")
    try:
        conn = sqlite3.connect("employee.db")

        name = input("Employee's name: ")
        position = input("Employee's position: ")
        salary = float(input("Employee's salary: "))

        query = conn.cursor()
        query.execute('INSERT INTO employees (name, position, salary) VALUES (?, ?, ?)', (name, position, salary))
        conn.commit()

        print("\nEmployee added successfully.")
    except sqlite3.Error:
        print("\nError connect database.")
    except ValueError:
        print("\nData invalid.")
    finally:
        conn.close()


def editEmployee():
    os.system('clear')
    figlet("Edit employee")
    try:
        conn = sqlite3.connect("employee.db")

        employeeId = int(input("Employee's Id: "))

        query = conn.cursor()
        query.execute('SELECT * FROM employees WHERE id = ?', (employeeId,))
        employee = query.fetchone()

        if employee:
            print(f"Employee found.\nEmployee: {employee[1]}, {employee[2]}, {employee[3]}")
            print("\nEnter new information (leave blank to keep current): ")
            newName = input(f"Enter new name: ")
            newPosition = input(f"Enter new position: ")
            newSalary = float(input(f"Enter new salary: "))

            newName = newName.strip() if newName.strip() else employee[1]
            newPosition = newPosition.strip() if newPosition.strip() else employee[2]
            newSalary = newSalary if newSalary else employee[3]

            updateQuery = 'UPDATE employees SET name = COALESCE(?, name), position = COALESCE(?, position), salary = COALESCE(?, salary) WHERE id = ?'
            query.execute(updateQuery, (newName, newPosition, newSalary, employeeId))
            conn.commit()

            print("\nEmployee information updated successfully.")
        else:
            print("\nEmployee not found.")
    except sqlite3.Error:
        print("Error connect database.")
    except ValueError:
        print("\nData invalid.")
    finally:
        conn.close()


def deleteEmployee():
    os.system('clear')
    figlet("Delete employee")
    try:
        conn = sqlite3.connect("employee.db")

        employeeId = int(input("Employee's Id: "))

        query = conn.cursor()
        query.execute('SELECT * FROM employees WHERE id = ?', (employeeId,))
        employee = query.fetchone()

        if employee:
            print(f"Employee found.\nEmployee: {employee[1]}, {employee[2]}, {employee[3]}")
            option = input("\nAre you sure you want to delete it? (y/n): ")

            if option == "y" or option == "Y":
                query.execute('DELETE FROM employees WHERE id = ?', (employeeId,))
                conn.commit()
                print("\nEmployee deleted successfully.")
        else:
            print("\nEmployee not found.")
    except sqlite3.Error:
        print("\nError connect database.")
    except ValueError:
        print("\nData invalid.")
    finally:
        conn.close()


def viewEmployee():
    os.system('clear')
    figlet("View employee")
    try:
        conn = sqlite3.connect("employee.db")
        query = conn.cursor()
        query.execute('SELECT * FROM employees')
        employees = query.fetchall()

        if employees:
            headers = ['ID', 'Name', 'Position', 'Salary']
            print(tabulate(employees, headers = headers, tablefmt = 'grid'))
        else:
            print("No employees found.")
    except sqlite3.Error:
        print("\nError connect database.")
    finally:
        conn.close()


def pause():
    input("Press Enter to continue...")


def figlet(text):
    figlet = Figlet()
    figlet = Figlet(font="mini")
    print(figlet.renderText(text))


def createDatabase():
    conn = sqlite3.connect("employee.db")
    query = conn.cursor()
    query.execute('CREATE TABLE IF NOT EXISTS employees (id INTEGER PRIMARY KEY, name TEXT NOT NULL, position TEXT NOT NULL, salary REAL NOT NULL)')
    conn.commit()
    conn.close()

if __name__ == "__main__":
    main()
